### PR TITLE
Simplify annotation and type passing in `BuilderMutatorFactory`

### PR DESCRIPTION
**Pull request by fmeum**

### CODE REVIEW

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/proto/BuilderMutatorFactory.java [hunk 1]
```diff
@@ -106,38 +110,69 @@ private static <T extends Builder, U> InPlaceMutator<T> mutatorForField(FieldDes
...
+      } else {
+        messageDescriptor = field.getMessageType();
       }
-    });
+      return new ChainedMutatorFactory(originalFactory, new MutatorFactory() {
+        @Override
+        public Optional<SerializingMutator<?>> tryCreate(
+            AnnotatedType type, MutatorFactory factory) {
+          return asSubclassOrEmpty(type, Message.Builder.class).flatMap(clazz -> {
+            if (clazz != Message.Builder.class) {
```
Comment (bertschneider) [inline src/main/java/com/code_intelligence/jazzer/mutation/mutator/proto/BuilderMutatorFactory.java:157]: I was somewhat confused why `originalFactary` wouldn't handle this type, as it's tried before. Could you add a comment an `builderClass [==|!=] Message.Builder.class` regarding this situation?

**Review by bertschneider — APPROVED** [review]
Nice simplification! Custom mutator factories turn out to be quite useful, again.

File: src/main/java/com/code_intelligence/jazzer/mutation/combinator/MutatorCombinators.java [hunk 2]
```diff
@@ -135,10 +135,8 @@ public String toString() {
       };
     }
 
+    final InPlaceMutator<T>[] mutators = Arrays.copyOf(partialMutators, partialMutators.length);
```
[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/code_intelligence/jazzer/mutation/combinator/MutatorCombinators.java:138]: The instance of the anonymous class used to retain direct references to both `partialMutators` and `mutators`, now it only retains `mutators`. Not a big difference, but there is little point in defensively copying an array if we also pass in the original one.

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/proto/BuilderAdapters.java [hunk 3]
```diff
@@ -174,10 +174,12 @@ static <T extends Builder, K, V> Map<K, V> getMapField(T builder, FieldDescripto
   static <T extends Builder, K, V> void setMapField(
       Builder builder, FieldDescriptor field, Map<K, V> map) {
     builder.clearField(field);
+    FieldDescriptor keyField = field.getMessageType().getFields().get(0);
+    FieldDescriptor valueField = field.getMessageType().getFields().get(1);
     for (Entry<K, V> entry : map.entrySet()) {
-      MapEntry.Builder<K, V> entryBuilder = (MapEntry.Builder) builder.newBuilderForField(field);
-      entryBuilder.setKey(entry.getKey());
-      entryBuilder.setValue(entry.getValue());
+      Builder entryBuilder = builder.newBuilderForField(field);
```
[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/code_intelligence/jazzer/mutation/mutator/proto/BuilderAdapters.java:180]: Nice optimization, will do.

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/proto/BuilderMutatorFactory.java (hunk 1 above)
Comment (fmeum) [reply src/main/java/com/code_intelligence/jazzer/mutation/mutator/proto/BuilderMutatorFactory.java:160]: Done!

**Pull request merged**